src/vm/hints/cairo_keccak/keccak_hints.rs

```diff
@@ -0,0 +1,414 @@
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let n_bytes = get_integer_from_var_name(
+        "n_bytes",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?;
+
+    let value = bigint!((n_bytes < &bigint!(BYTES_IN_WORD)) as usize);
+
+    vm.memory
+        .insert(&vm.run_context.ap, &MaybeRelocatable::from(value))
+        .map_err(VirtualMachineError::MemoryError)
+}
+
+/*
+Implements hint:
+    Cairo code:
+    if nondet %{ ids.n_bytes >= ids.KECCAK_FULL_RATE_IN_BYTES %} != 0:
+
+    Compiled code:
+    "memory[ap] = to_felt_or_relocatable(ids.n_bytes >= ids.KECCAK_FULL_RATE_IN_BYTES)"
+*/
+pub fn compare_keccak_full_rate_in_bytes_nondet(
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let n_bytes = get_integer_from_var_name(
+        "n_bytes",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?;
+
+    let value = bigint!((n_bytes >= &bigint!(KECCAK_FULL_RATE_IN_BYTES)) as usize);
+
+    vm.memory
+        .insert(&vm.run_context.ap, &MaybeRelocatable::from(value))
+        .map_err(VirtualMachineError::MemoryError)
+}
+
+/*
+Implements hint:
+    %{
+        from starkware.cairo.common.cairo_keccak.keccak_utils import keccak_func
+        _keccak_state_size_felts = int(ids.KECCAK_STATE_SIZE_FELTS)
+        assert 0 <= _keccak_state_size_felts < 100
+
+        output_values = keccak_func(memory.get_range(
+            ids.keccak_ptr - _keccak_state_size_felts, _keccak_state_size_felts))
+        segments.write_arg(ids.keccak_ptr, output_values)
+    %}
+*/
+
+pub fn block_permutation(
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let keccak_ptr = get_ptr_from_var_name(
+        "keccak_ptr",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?;
+
+    if KECCAK_STATE_SIZE_FELTS >= 100 {
+        return Err(VirtualMachineError::InvalidKeccakStateSizeFelts(
+            KECCAK_STATE_SIZE_FELTS,
+        ));
+    }
+
+    let values = vm
+        .memory
+        .get_range(
+            &MaybeRelocatable::RelocatableValue(keccak_ptr.sub(KECCAK_STATE_SIZE_FELTS)?),
+            KECCAK_STATE_SIZE_FELTS,
+        )
+        .map_err(VirtualMachineError::MemoryError)?;
+
+    let mut u64_values = maybe_reloc_vec_to_u64_array(&values)?;
+
+    // this function of the keccak crate is the one used instead of keccak_func from
+    // keccak_utils.py
+    keccak::f1600(&mut u64_values);
+
+    let bigint_values = u64_array_to_bigint_vec(&u64_values);
+
+    vm.segments
+        .write_arg(&mut vm.memory, &keccak_ptr, &bigint_values, Some(&vm.prime))
```

Comment:
Is there any chance to pass `bigint_values` as an iterator instead? This would save us from actually creating the vector.